File: js/src/lib/journeys/accountJourneys.ts
import { Page } from '@playwright/test';
import { TestUser } from '../fixtures/testUsers.js';
import { JourneyId, journeyIds, journeySpecs } from './generatedJourneySpecs.js';
import { executeJourney } from './journeyRunner.js';

export interface JourneyResult {
  journeyId: string;
  success: boolean;
  details?: string;
}

export type Journey = (page: Page, user: TestUser) => Promise<JourneyResult>;

export const runJourney = async (page: Page, user: TestUser, journeyId: JourneyId): Promise<JourneyResult> => {
  await executeJourney(page, user, journeyId);
  return { journeyId, success: true, details: journeySpecs[journeyId].successMessage };
};

export const loginAndViewDashboardJourney = async (page: Page, user: TestUser): Promise<JourneyResult> => {
  return runJourney(page, user, 'login-and-view-dashboard');
};

export const viewAccountDetailsJourney = async (page: Page, user: TestUser): Promise<JourneyResult> => {
  return runJourney(page, user, 'view-account-details');
};

export const journeys = Object.fromEntries(
  journeyIds.map((journeyId) => [journeyId, (page: Page, user: TestUser) => runJourney(page, user, journeyId)])
) as Record<JourneyId, Journey>;

File: py/src/playwright_adventures/journeys/account_journeys.py
from playwright.async_api import Page

from .generated_specs import JOURNEY_SPECS, JourneyId
from .journey_runner import execute_journey
from .models import JourneyResult, TestUser, demo_user


async def run_journey(page: Page, journey_id: JourneyId, user: TestUser = demo_user) -> JourneyResult:
    await execute_journey(page, user, journey_id)
    return JourneyResult(journey_id=journey_id, success=True, details=JOURNEY_SPECS[journey_id]["successMessage"])


async def login_and_view_dashboard(page: Page, user: TestUser = demo_user) -> JourneyResult:
    return await run_journey(page, "login-and-view-dashboard", user)


async def view_account_details(page: Page, user: TestUser = demo_user) -> JourneyResult:
    return await run_journey(page, "view-account-details", user)
